pilot/pkg/proxy/envoy/v1/watcher.go

```diff
@@ -55,34 +56,47 @@ type CertSource struct {
 }
 
 type watcher struct {
-	agent    proxy.Agent
-	role     model.Node
-	config   meshconfig.ProxyConfig
-	certs    []CertSource
-	pilotSAN []string
+	agent         proxy.Agent
+	role          model.Node
+	config        meshconfig.ProxyConfig
+	requiredCerts []CertSource // These certs must present for Envoy to start
+	certs         []CertSource
+	pilotSAN      []string
 }
 
 // NewWatcher creates a new watcher instance from a proxy agent and a set of monitored certificate paths
 // (directories with files in them)
 func NewWatcher(config meshconfig.ProxyConfig, agent proxy.Agent, role model.Node,
-	certs []CertSource, pilotSAN []string) Watcher {
+	internalCerts, externalCerts []CertSource, pilotSAN []string) Watcher {
 	return &watcher{
-		agent:    agent,
-		role:     role,
-		config:   config,
-		certs:    certs,
-		pilotSAN: pilotSAN,
+		agent:         agent,
+		role:          role,
+		config:        config,
+		requiredCerts: internalCerts,
+		certs:         append(internalCerts, externalCerts...),
+		pilotSAN:      pilotSAN,
 	}
 }
 
 const (
+	requiredCertsCheckInterval = 4 * time.Second
+	requiredCertsCheckTimeout  = 90 * time.Second
 	// defaultMinDelay is the minimum amount of time between delivery of two successive events via updateFunc.
 	defaultMinDelay = 10 * time.Second
 	azRetryInterval = time.Second * 30
 	azRetryAttempts = 10
 )
 
 func (w *watcher) Run(ctx context.Context) {
+	// Wait until the required certificate files present. This prevents Envoy failing caused by non-existing cert files.
+	// Note: this function is blocking.
+	log.Infof("Check required cert files are present (up to %v)...", requiredCertsCheckTimeout)
+	if err := waitForCertsPresent(w.requiredCerts, requiredCertsCheckInterval, requiredCertsCheckTimeout); err != nil {
+		log.Errorf("Required cert file check failed: %v", err)
```

Comment:
How is this error propagated upward? Does the caller fallback to non-mtls mode or have some other graceful recovery mode?

Reply:
I don't think we have a graceful recovery mode, this PR only fixes the issue that Envoy is started before the cert files are ready. In case of a timeout, I think the node agent will exit and kubelet will try to restart the node agent. Fallback to non-mtls mode does not work because the service is discovered as mTLS service, unless it can somehow change its discovery info.

Reply:
I'm not sure how the node agent would exit if we don't return an error here (or crash/fatal). Does envoy start-up eventually abort if certs aren't present?

Reply:
The line 96 `return` make the node agent exit, right? I will add `<-ctx.Done()` before `return` to make sure the context is done.
Envoy will never start if the certs are not ready.

Reply:
Is the node-agent part of the pilot-agent or a separate process?

Reply:
Can you confirm kubelet will restart by not installing istio-CA and not having the certificates?

Reply:
Sorry, I meant the pilot agent in the previous comment. I have updated it.

Reply:
> The line 96 return make the pilot agent exit, right? 

I don't see how. `watcher.Run` is started as a goroutine. Nothing in this function feedsback errors/status to the caller function. You'd have to pass a cancel channel/function (via context.WithCancel), which would be invoked on error, and then make the caller monitor `ctx.Done()` to handle the error condition.

Reply:
Thanks @ayj, I think we can use another channel to notify the pilot agent main function that the watcher has failed. The main function may terminate, run the watcher again, or do some recovery (this work is TBD).

Reply:
How do you know if mTLS is enabled or not? Also what about incremental mTLS enablement where pod is already up and running ?